common/src/main/java/tachyon/worker/WorkerClient.java

```diff
@@ -114,101 +108,106 @@ public synchronized void accessBlock(long blockId) throws IOException {
-    mustConnect();
-
-    try {
-      return mClient.asyncCheckpoint(fileId);
-    } catch (TachyonTException e) {
-      throw new IOException(e);
-    } catch (TException e) {
-      mConnected = false;
-      throw new IOException(e);
-    }
+  public synchronized boolean asyncCheckpoint(final long fileId) throws IOException,
+      TachyonException {
+    return retryRPC(new RpcCallableThrowsTachyonTException<Boolean>() {
+      @Override
+      public Boolean call() throws TachyonTException, TException {
+        return mClient.asyncCheckpoint(fileId);
+      }
+    });
   }
 
   /**
    * Notifies the worker the block is cached.
    *
    * @param blockId The id of the block
-   * @throws IOException
+   * @throws IOException if an I/O error occurs
+   * @throws TachyonException if a Tachyon error occurs
    */
-  public synchronized void cacheBlock(long blockId) throws IOException {
-    mustConnect();
-
-    try {
-      mClient.cacheBlock(mSessionId, blockId);
-    } catch (TachyonTException e) {
-      throw new IOException(e);
-    } catch (TException e) {
-      mConnected = false;
-      throw new IOException(e);
-    }
+  public synchronized void cacheBlock(final long blockId) throws IOException, TachyonException {
+    retryRPC(new RpcCallableThrowsTachyonTException<Void>() {
+      @Override
+      public Void call() throws TachyonTException, TException {
+        mClient.cacheBlock(mSessionId, blockId);
+        return null;
+      }
+    });
   }
 
   /**
    * Notifies worker that the block has been cancelled
    *
    * @param blockId The Id of the block to be cancelled
-   * @throws IOException
+   * @throws IOException if an I/O error occurs
+   * @throws TachyonException if a Tachyon error occurs
    */
-  public synchronized void cancelBlock(long blockId) throws IOException {
-    mustConnect();
-
-    try {
-      mClient.cancelBlock(mSessionId, blockId);
-    } catch (TException e) {
-      mConnected = false;
-      throw new IOException(e);
-    }
+  public synchronized void cancelBlock(final long blockId) throws IOException, TachyonException {
+    retryRPC(new RpcCallableThrowsTachyonTException<Void>() {
+      @Override
+      public Void call() throws TachyonTException, TException {
+        mClient.cancelBlock(mSessionId, blockId);
+        return null;
+      }
+    });
   }
 
   /**
    * Closes the connection to worker. Shutdown the heartbeat thread.
    */
   @Override
-  public synchronized void close() {
-    if (mConnected) {
-      try {
-        // Heartbeat to send the client metrics.
-        if (mHeartbeatExecutor != null) {
-          mHeartbeatExecutor.heartbeat();
-        }
-        mProtocol.getTransport().close();
-      } finally {
-        if (mHeartbeat != null) {
-          mHeartbeat.cancel(true);
-        }
-      }
-      mConnected = false;
+  protected void closeOperation() {
+    // Heartbeat to send the client metrics.
```

Comment:
As I mentioned in `ClientBase`, if we override `beforeDisconnect` here, we don't need to call method from `super`.